### PR TITLE
 Clean up code to determine line items for membership batch entry

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
 Clean up code to determine line items for membership batch entry

Before
----------------------------------------
whoa

After
----------------------------------------
Whee

Technical Details
----------------------------------------
This is towards switching this code to use the order api as it currently uses the BAO to create it's contributions

Comments
----------------------------------------
Builds on #20778

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I suspect this is the underlying e-notice - closing until I've fixed

Stacktrace
CRM_Member_Form_MembershipTest::testSubmitRecur
CiviCRM_API3_Exception: Illegal string offset 'line_total'

/home/jenkins/bknix-dfl/build/core-20779-4u7oy/web/sites/all/modules/civicrm/api/api.php:134
/home/jenkins/bknix-dfl/build/core-20779-4u7oy/web/sites/all/modules/civicrm/CRM/Member/Form/Membership.php:1153
/home/jenkins/bknix-dfl/build/core-20779-4u7oy/web/sites/all/modules/civicrm/CRM/Member/Form.php:526
/home/jenkins/bknix-dfl/build/core-20779-4u7oy/web/sites/all/modules/civicrm/tests/phpunit/CRM/Member/Form/MembershipTest.php:795
/home/jenkins/bknix-dfl/build/core-20779-4u7oy/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:256
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671

**Comment by colemanw** [comment]
@eileenmcnaughton I'm not sure why the commit from #20778 is still showing up in this PR. Can you try rebasing?

**Comment by eileenmcnaughton** [comment]
@colemanw done

File: CRM/Financial/BAO/Order.php [hunk 1]
```diff
@@ -742,22 +760,45 @@ public function getLineItem($index): array {
    */
   protected function fillMembershipLine(array $lineItem): array {
     $fields = $this->getPriceFieldsMetadata();
-    $field = reset($fields);
```
Comment (eileenmcnaughton) [inline CRM/Financial/BAO/Order.php:745]: this incorrectly assumed there would only be one possible field - but it's one per membership organization (the field in the membership_type table)

File: CRM/Financial/BAO/Order.php [hunk 2]
```diff
@@ -742,22 +760,45 @@ public function getLineItem($index): array {
...
 
+  /**
+   * Add total_amount and tax_amount to the line from the override total.
+   *
+   * @param int $financialTypeID
+   * @param array $lineItem
+   *
+   * @return void
+   */
+  protected function addTotalsToLineBasedOnOverrideTotal(int $financialTypeID, array &$lineItem): void {
```
Comment (eileenmcnaughton) [inline CRM/Financial/BAO/Order.php:792]: extraction

**Comment by eileenmcnaughton** [comment]
@colemanw are you comfortable merging this - it's less scary & more tested than it looks - but I agree it looks a bit scary :-)

**Comment by colemanw** [comment]
Ok I see this is mostly code cleanup and test coverage is good.

**Comment by eileenmcnaughton** [comment]
@colemanw yeah - it was test cover that made me realise this code is doing something weird :-)

**Pull request merged**